Added support for `Ctr+Enter` to complete text input in Text tool dialog

mohamedAdhamc opened the pull request:
Fix #2041 
Hey so I have been trying to do this for a bit, I started first with trying to implement accelerators but then realized that they are more for "menu" related buttons and such. So I was thinking of catching keypress events and then if the correct key is pressed then the text will be applied. However, in the process of implementing i'm facing multiple isssues.

1- `signal_key_press_event().connect(sigc::ptr_fun(&on_key_pressed)); `    in all other places where this is implemented its always done in the constructor of the class related to it. for example for dialog-preview it is in its constructor. also in gtkmm documentation it mentions it being placed in the constructor. by placing it here the build gives out an error :'signal_key_press_event' was not declared in this scope
the problem here is that dialog-paragraph is a method not a class so I cant really place it in its constructor. to which I have thought of trying to place it instead in the app class constructor (since it is the class which includes this method)... still to no successs

I'm getting puzzled here as to what I could do or where to place it.

2-also for the line 3785
 I'm getting two errors "expected primary-expression before '*' token" and "ev was not declared in this scope"


p.s
the body of "on_key_pressed" is not really made at all I'm focusing now on how to actually apply this method.

ice0 commented:
> 1- `signal_key_press_event().connect(sigc::ptr_fun(&on_key_pressed)); ` in all other places where this is implemented its always done in the constructor of the class related to it. for example for dialog-preview it is in its constructor.

You can attach to event in any place.
In your case this will be `<widget>.signal_key_press_event().connect(...);`.
For example `dialog.signal_key_press_event().connect`.

> also in gtkmm documentation it mentions it being placed in the constructor. by placing it here the build gives out an error :'signal_key_press_event' was not declared in this scope the problem here is that dialog-paragraph is a method not a class so I cant really place it in its constructor. to which I have thought of trying to place it instead in the app class constructor (since it is the class which includes this method)... still to no successs

I think it's better to start from simple test application, so you will not be confused with unrelated stuff.

> 
> I'm getting puzzled here as to what I could do or where to place it.
> 
> 2-also for the line 3785 I'm getting two errors "expected primary-expression before '*' token" and "ev was not declared in this scope"
> 
> p.s the body of "on_key_pressed" is not really made at all I'm focusing now on how to actually apply this method.

It looks like you are trying to make a function inside a function. Please check this discussion for better understanding: https://stackoverflow.com/questions/4324763/can-we-have-functions-inside-functions-in-c

mohamedAdhamc commented:
Hey, @ice0 thanks for the head up. now the key-press signal is being handled correctly I believe as when the dialog is open and i hit ctr+enter it prints out "hey" in the command prompt.

I believe I'm almost done here the last thing would be to implement line 3772 `//ok_button->clicked();`
so then when the shortcut is hit the button will be clicked. However, I'm having quite a problem here trying to access the button which is the dialog_paragraph method from the on_key_pressed method.  any suggestions as to what i might do ?

p.s, a few side questions
1- as you might have noticed I have made an entirely new button xd just so that I can have its name as I couldn't find in the documentation when dialog.add_button is used if "ok" is the name or if it even has one. any suggestions?

2- also I found that there are three key values for enter,  is that related to each machine type, and if so should I keep all three as is ? (GDK_KEY_KP_Enter, GDK_KEY_ISO_Enter, GDK_KEY_3270_Enter)

ice0 commented:
> now the key-press signal is being handled correctly I believe as when the dialog is open and i hit ctr+enter it prints out "hey" in the command prompt.

I'm not sure about this because the `ev->keyval` check is wrong.
Check the LGTM warning first to see what I mean. 

> I believe I'm almost done here the last thing would be to implement line 3772 `//ok_button->clicked();` so then when the shortcut is hit the button will be clicked. However, I'm having quite a problem here trying to access the button which is the dialog_paragraph method from the on_key_pressed method. any suggestions as to what i might do ?

I think there is no need to "click" the button. This the quote from documentation (https://docs.gtk.org/gtk3/class.Dialog.html):

>  When adding buttons using add_button(), clicking the button will emit
 signal_response() with a "response id" you specified. You are encouraged
 to use the Gtk::ResponseType enum.

It's better to call the `Gtk::Dialog::response(...)` method with the required `response_id` to emit required signal. 


> p.s, a few side questions 1- as you might have noticed I have made an entirely new button xd just so that I can have its name as I couldn't find in the documentation when dialog.add_button is used if "ok" is the name or if it even has one. any suggestions?

If you need the button which has `RESPONSE_OK` you can use `get_widget_for_response(...)` method.
But, as I mentioned earlier, there is no reason to do this, because you can just emit required signal. 

> 
> 2- also I found that there are three key values for enter, is that related to each machine type, and if so should I keep all three as is ? (GDK_KEY_KP_Enter, GDK_KEY_ISO_Enter, GDK_KEY_3270_Enter)

You need 2 values: `GDK_KEY_Return` and `GDK_KEY_KP_Enter`
Return is "Enter" on the main keyboard and "KP_Enter" on the keypad/numpad.

mohamedAdhamc commented:
> 
> I'm not sure about this because the `ev->keyval` check is wrong. Check the LGTM warning first to see what I mean.
> 
> > I believe I'm almost done here the last thing would be to implement line 3772 `//ok_button->clicked();` so then when the shortcut is hit the button will be clicked. However, I'm having quite a problem here trying to access the button which is the dialog_paragraph method from the on_key_pressed method. any suggestions as to what i might do ?
> 


@ice0 yes, I can how using a logical operator here in that manner could possibly cause an error. however, it suggests using bitwise operators which would just output a whole new value which isn't needed I believe. However, I don't believe that's much of an issue, worst case scenario would be having to split it into two if-conditions.

Now, the bigger problem and which is causing more of an issue is that enter isn't being detected. I put a `std::cout<<(ev->keyval);`  at the very beginning of the method called when a key is pressed. I tried different combinations and the only ones  that seem to be picked are (control+letter/numbers), letters by themselves aren't, also enter, and basically anything which actually manipulates the text box it seems. could this actually be the reason ?

ice0 commented:
> Now, the bigger problem and which is causing more of an issue is that enter isn't being detected. I put a `std::cout<<(ev->keyval);` at the very beginning of the method called when a key is pressed. 

There is a possibility that the keystroke is intercepted somewhere else. For example, in the main Synfig window.
And that's exactly the reason why I recommended that you start with a simple application to make sure that there are no other factors influencing the behavior. 

> I think it's better to start from simple test application, so you will not be confused with unrelated stuff.

rodolforg commented:
The signal_key_press_event must be attached to the text area, not to the OK/Cancel buttons.

mohamedAdhamc commented:
> The signal_key_press_event must be attached to the text area, not to the OK/Cancel buttons.

yes, I had it attached at the beginning to the dialog not the buttons, but for some reason I thought the shortcut applying the button would look better to the eye if it was to be noticed, but yea as ice said using the dialog response right away is better.

 The problem I seem to be facing now though is that I cant access the dialog which is in the paragraph layer function from the on_key_press function. Also for some reason ctr+ enter/up/down/left or right isn't captured and I'm suspecting it has to do with the text buffer maybe picking the signal first or something as also letters by themselves aren't picked up they're only picked up when ctrl is pressed and the text cursor kind of pauses. I was looking through the possible solutions but was quite busy in the past couple days I will hopefully be back at it tommorow. Any suggestions would be appreciated though  😊 .

mohamedAdhamc commented:
quick description of what I have done so far:

1- `dialog.signal_key_press_event().connect(sigc::ptr_fun(&on_key_pressed));`   for catching the key-press event

2- then the on_key_press method which sets a variable to one 

3- supposedly when this variable is 1 a gtk response ok signal is emitted which the run() handles by default.

4- at the end of the dialog execution the variable value is set to zero again.

what I'm having trouble with though I believe is step 3. From what I read the run function executes a recursive main loop until a button is pressed or a response signal is emitted then when either of them happens it returns the response id and the program flow continues normally. I even tried directly harcoding an ok response right before the .run() and still it didnt return anything ! any ideas on what I can do ?

p.s I was using Github desktop and it ruined the files I will be adding the correct commits later on today as I have to go for now.

ice0 commented:
> what I'm having trouble with though I believe is step 3.

I'm not sure about this. I made a simple application as I already recommended to you and found the reason why the keypress is not being handled correctly. After that, I switched to Synfig and implemented the necessary logic.
I used a lambda instead of the explicit `on_key_pressed` function just to keep all the logic in one place.
The whole solution consists of only 9 lines of code.

mohamedAdhamc commented:
> > what I'm having trouble with though I believe is step 3.
> 
> I'm not sure about this. I made a simple application as I already recommended to you and found the reason why the keypress is not being handled correctly. After that, I switched to Synfig and implemented the necessary logic. I used a lambda instead of the explicit `on_key_pressed` function just to keep all the logic in one place. The whole solution consists of only 9 lines of code.

I tried the simple application method, but I found that when I mimic the exact behaviour of this dialog I get the same problems, so I though then maybe doing it in the same place would save time ().

ice0 commented:
> However, now since you have already done it should I close this PR ?

No. I just want to say that this problem has a simple solution and there is a good chance to learn something new here.

mohamedAdhamc commented:
> > However, now since you have already done it should I close this PR ?
> 
> No. I just want to say that this problem has a simple solution and there is a good chance to learn something new here.

Oh alright then should I continue working on it ?

ice0 commented:
> Oh alright, then I should keep working on it correct ?

Sure.

mohamedAdhamc commented:
Unfortunately due to my exams, which will be everyday in the upcoming week starting saturday, I will have to take a break for this week. it's been a good couple weeks now with you guys. Looking forward to being back after a week with even more enthusiasm to continue this and other new things 😊.

mohamedAdhamc commented:
Hey all, I'm finally back 😊
so the lambda seems to be implemented correctly. Also ctrl+s succesfully applies a gtk response ok. However, still working on making ctrl+enter work.

 when I made a simple dialog without the text view ctrl+ enter worked just fine. Also when I did the same for synfig by commenting text_view.show(); it worked correctly (the signal was attached to dialog not text_view while doing so). I then also tried using  text_view.set_editable(false); on the condition that control is pressed which also makes ctrl+enter works (but obviously messses up other things). So basically I'm 99% sure the problem here is from the text view/ text buffer as whenever a key is pressed which manipulates them it's handled right away by the text_view/buffer before being able to be handled by the lambda.

I have been going through the docs trying to find out a good way of overriding this behaviour only for ctr+enter. I found multiple things like `im_context_filter_keypress( )` , and I tried to implement them but I'm also not sure if this is the "correct" way there might be an even easier way for overriding this default behaviour. I guess thats all for today tommrow i will be continuing hopefully.

mohamedAdhamc commented:
@ice0  done. now ctr+enter/return works correctly to apply change in the dialog paragraph. This alongside #2582  should finally fix #2041.

ice0 commented on 2-build-msys-cmake.sh:
```diff
@@ -3,7 +3,7 @@
 export PKG_CONFIG_PATH="${PKG_CONFIG_PATH}:/opt/mlt-7.2.0/lib/pkgconfig"
 mkdir -p cmake-build-msys && cd cmake-build-msys
 mode="Release"
-if [[ -n $1 ]] 
```
You need to remove this change.

ice0 commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -3777,6 +3778,22 @@ App::dialog_paragraph(const std::string &title, const std::string &message,std::
+				(ev->keyval == (GDK_KEY_Return) || ev->keyval == (GDK_KEY_KP_Enter) ) &&
+				(ev->state == GDK_CONTROL_MASK) )
+			{
+			dialog.response(-5);
```
Please use a named constant here (instead of -5).

mohamedAdhamc replied:
oh yes definetly xd, was planning on doing so at the end but forgot.

ice0 commented:
I suggest to copy changes from https://github.com/synfig/synfig/pull/2582 here

mohamedAdhamc commented:
@ice0 done.

rodolforg commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -3773,10 +3774,26 @@ App::dialog_paragraph(const std::string &title, const std::string &message,std::
+	text_view.signal_key_press_event().connect(( [&](GdkEventKey *ev) {
+
+		SYNFIG_EXCEPTION_GUARD_BEGIN()
+		if (	(ev->type == GDK_KEY_PRESS) &&
+				(ev->keyval == (GDK_KEY_Return) || ev->keyval == (GDK_KEY_KP_Enter) ) &&
+				(ev->state == GDK_CONTROL_MASK) )
+			{
+			dialog.response(Gtk::RESPONSE_OK);
+			return true;
+			}
+			return false;
```
Indentation is a bit weird here :)

mohamedAdhamc replied:
yea, I admit it did look a bit weirx xd. I fixed it up abit and indented the  "if statement" in the way you guys proposed in the forums. I am guilty of using a space or two though to fix the allignment of the multi-lined condition but I think its fine. if not please let me know 😊.

rodolforg commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -3773,10 +3774,25 @@ App::dialog_paragraph(const std::string &title, const std::string &message,std::
+		SYNFIG_EXCEPTION_GUARD_END_BOOL(true)
+
+	}), false );
+
+
```
No need for those 2 extra blank lines

rodolforg commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -3773,10 +3774,25 @@ App::dialog_paragraph(const std::string &title, const std::string &message,std::
+	dialog.add_button(_("_OK"),   Gtk::RESPONSE_OK)->set_image_from_icon_name("gtk-ok", Gtk::ICON_SIZE_BUTTON);
 	dialog.set_default_response(Gtk::RESPONSE_OK);
 
+	text_view.signal_key_press_event().connect(( [&](GdkEventKey *ev) {
```
You don't have to place all App object in lambda context (And we should avoid some lambda context usages).
Use only the needed variable: `[&dialog]`.

mohamedAdhamc replied:
yes makes much more sense to not include everything in the laambda. Thanks !

mohamedAdhamc commented:
damn, I tried rebasing using github desktop and now lots of thing got messede up (* _ *)

ice0 commented:
Merged via #2689
